### PR TITLE
build: enhance dependency management and make `scipy` optional.

### PR DESCRIPTION
**Pull request by egparedes**
Added:
- New requirements files for oldest versions and oldest versions with extras.
- Documentation for the semi-automatic dependency management workflow with pip-tools and cog.
- Daily CI badge in the main README file.
- New `ufuncs` module for NativeFunctions.

Changed:
- Update and fix minimum gt4py package dependencies.
- Update and fix minimum development dependencies.
- Rename `docs/development/CI` folder to `docs/development/tools`.
- Small additions to ci-infrastructure documentation.
- Daily CI action now tests several requirements files: more recent, oldest versions and oldest versions with extras.
- Small enhancements to tox.ini, which now uses optional environment variables for overriding some settings like the requirements files used for the test environment.
- Refactor ufuncs and code generation for cartesian-NumPy backend.

### CODE REVIEW

**Comment by havogt** [comment]
It's a simple change, but with quite heavy maintenance cost, unless we find an easy way to test both cases (scipy, no scipy).

File: .github/workflows/daily-ci.yml [hunk 1]
```diff
@@ -5,9 +5,9 @@ on:
   - cron: '0 4 * * *'
 
   ## COMMENTED OUT: only for testing CI action changes
-  ## pull_request:
-  ##   branches:
-  ##   - main
+  pull_request:
+    branches:
+    - main
```
Comment (havogt) [inline .github/workflows/daily-ci.yml:10]: Just a reminder to change back before merge.

**Review by havogt — APPROVED** [review]
Looks good to me.

**Review by gronerl — APPROVED** [review]

**Review by DropD — APPROVED** [review]
Docs are ok.

**Pull request merged**